src/sc_stage.py

```diff
@@ -303,56 +322,104 @@ def mark_item_delivered(self, message: Message):
         '''
-        if self.item_dep.has_prereq():
-            try:
-                # For each manufactured batch item, do a get() in the inbound queue once.
-                # NOTE/FIXME/TODO
-                # This does not work if batches of different item quantities
-                # and item types are required by the stage's inbound requirements.
-                # The manufacture logic assumes each result batch can be made with
-                # any, single, incoming batch.
-                log.info("Node %d waiting for %s from %s", self.node_id, self.item_dep.get_prereq(), suppliers)
-                material_req = self.inbound_material.get(timeout=5)  # HACK wait at most 5 seconds for a material batch
-                self.inbound_log[material_req] = StageStatus.CONSUMED
-                log.info("Node %d successfully consumed %s", self.node_id, material_req)
-            except Empty:
-                # queue is empty, need to wait till it's not.
-                log.warning("Node %d's inbound material queue is empty but needs %s. Skipping manufacturing of one batch of item type %s",
-                            self.node_id, self.item_dep.get_prereq(), self.get_stage_result_type())
+            suppliers: list of (node_id, item_req) tuples obtained from the latest flow
+        '''
+        prereqs = self.item_dep.get_prereq()
+        if prereqs:
+            queue_types = set(self.inbound_material.keys())
+            supplier_types = set([ir.item.type for _, ir in suppliers])
+
+            suppliers = {ir.item.type: sid for sid, ir in suppliers}  # convert the flow edges to dict
+
+            if not queue_types.issuperset(supplier_types):  # verify there is queue for each supplier
+                log.critical("Node {} has more suppliers ({}) from flow the available queue types ({})"
+                             .format(self.node_id, supplier_types, queue_types))
+                log.error("Node %d skipping manufacturing cycle", self.node_id)
+                return
+
+            has_all_materials = True
+            for item_type, supplier_id in suppliers.items():
+                queue = self.inbound_material[item_type]
+
+                if queue.empty():
+                    log.info("Node %d needs batch of type %s from node %s", self.node_id, item_type, supplier_id)
+                    self._send_material_request_upstream(supplier_id, item_type)
+                    has_all_materials = False
+
+            if has_all_materials:
+                for item_type, supplier_id in suppliers.items():
+                    queue = self.inbound_material[item_type]
+
+                    material_req = queue.get(timeout=1)  # HACK wait at most 1 second for a material batch
+                    self.inbound_log[material_req] = BatchStatus.CONSUMED
+                    log.info("Node %d successfully consumed %s", self.node_id, material_req)
+                    self.consumed_count += 1
```

Comment:
Add consumption count at item type granularity

Reply:
consumed and manufactured counts are for batches not individual items.